src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -461,21 +433,14 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
                     _sslState.FinishWrite();
 
                 } while (count != 0);
-            }
 
-            if (asyncRequest != null) 
-            {
-                asyncRequest.CompleteUser();
             }
+            asyncRequest?.CompleteUser();
```

Comment:
CompleteUser is basically just calling the user callback.  There shouldn't be any contention in the CompleteUser call itself (though of course, there could be).

The main difference here is likely something in the user code.  To test this, you could add Task.Run to the user code so that the user callback will be enqueued, instead of run immediately, and see if that makes any difference.

Which test are you seeing this with?  I'm not sure quite what the user's callback will be doing in this case.